[DOCS] Add data stream overview and intro (#57596)
elastic/elasticsearch: docs/reference/data-streams.asciidoc

This new index will then become the new write index.

[float]
[[getting-started-with-data-streams]]
== Getting started with data streams

Create a composable template with a data stream definition:

[source,console]
-----------------------------------
PUT /_index_template/logs_template
{
  "index_patterns": ["logs-*"],
  "data_stream": {
    "timestamp_field": "@timestamp"
  }
}
-----------------------------------
// TEST

Start indexing data into the namespace of this composable template:

[source,console]
--------------------------------------------------
POST /logs-foobar/_doc
--------------------------------------------------
// TEST[continued]

Response:

[source,console-result]
--------------------------------------------------
{
    "_shards" : {
        "total" : 1,
        "failed" : 0,
        "successful" : 1
    },
    "_index" : "logs-foobar-000001",
    "_type" : "_doc",
    "_id" : "W0tpsmIBdwcYyG50zbta",
    "_version" : 1,
    "_seq_no" : 0,
    "_primary_term" : 1,
    "result": "created"
}
--------------------------------------------------
// TESTRESPONSE[s/W0tpsmIBdwcYyG50zbta/$body._id/

Or create a data stream via the create data stream api:

> [source,console-result]

Reviewer: ```suggestion
[source,console]
```

[[data-streams]]
= Data Streams

[partintro]
--
A data stream formalizes the notion of a stream of data for time based data sources.

> Data streams groups indices from the same time-based data source together as an

Reviewer 2: ```suggestion
Data streams group indices from the same time-based data source together as an
```

> opaque container. A data stream keeps track of an ordered list of indices. Also known

Reviewer 2: ```suggestion
opaque container. A data stream keeps track of an ordered list of indices which are also known
```

as backing indices.

The backing indices of a data steam are hidden indices. The last backing index is the

> write index. The other backing indices handle read requests. A data steam only accepts

Reviewer 2: ```suggestion
write index. The other backing indices handle read requests. A data stream only accepts
```

index requests with `op_type` set to `create`. If specific documents need to be updated
or deleted then these operations should happen via the backing index these documents
reside in.

Data streams rely on composable templates in order to get created. A composable template

> with a `data_stream` definition must exists for the namespace the data stream gets
> created in. A data stream is created automatically when indexing into namespace with

Reviewer 2: ```suggestion
with a `data_stream` definition must exist for the namespace in which the data stream
is to be created. A data stream is created automatically when indexing into a namespace with
```

a composable template with a `data_stream` definition. A data stream can also

> be created via the Create Data Stream API, but this also requires the existence of

Reviewer 2: ```suggestion
be created via the <<indices-create-data-stream,Create Data Stream API>>, but this also requires the existence of
```

[source,console]
--------------------------------------------------
POST /logs-foobar/_doc
--------------------------------------------------
// TEST[continued]

Response:

[source,console-result]
--------------------------------------------------
{
    "_shards" : {
        "total" : 1,
        "failed" : 0,
        "successful" : 1
    },
    "_index" : "logs-foobar-000001",
    "_type" : "_doc",
    "_id" : "W0tpsmIBdwcYyG50zbta",
    "_version" : 1,
    "_seq_no" : 0,
    "_primary_term" : 1,
    "result": "created"
}
--------------------------------------------------
// TESTRESPONSE[s/W0tpsmIBdwcYyG50zbta/$body._id/

Or create a data stream via the create data stream api:

[source,console-result]
--------------------------------------------------
PUT /_data_stream/logs-barbaz
--------------------------------------------------
// TEST[continued]

////
[source,console]
-----------------------------------
DELETE /_data_stream/logs-foobar
DELETE /_data_stream/logs-barbaz
DELETE /_index_template/logs_template
-----------------------------------
// TEST[continued]
////

[float]
[[data-streams-apis]]
== Data stream APIs

> The following APIs are for managing data stream APIs:

Reviewer 2: ```suggestion
The following APIs are available for managing data streams:
```

> * To get meta information about data streams use <<indices-get-data-stream, the get data stream api>>.

Reviewer 2: ```suggestion
* To get information about data streams, use the <<indices-get-data-stream, get data stream API>>.
```

> * To delete data streams use <<indices-delete-data-stream, the delete data stream api>>.

Reviewer 2: ```suggestion
* To delete data streams, use the <<indices-delete-data-stream, delete data stream API>>.
```

> * To manually create a data stream use <<indices-create-data-stream, the create data stream api>>.

Reviewer 2: ```suggestion
* To manually create a data stream, use the <<indices-create-data-stream, create data stream API>>.
```

in the time-based data source should be configured with the data stream. This can be
specified in the `timestamp_field` parameter in the `data_stream` snippet of a
composable template.

The rollover API is responsible for updating <<rollover-data-stream-ex,a data stream when rolling over>>.
It will create a new index and update the data stream to include it the list of indices.
This new index will then become the new write index.

[float]
[[getting-started-with-data-streams]]
== Getting started with data streams

Create a composable template with a data stream definition:

[source,console]
-----------------------------------
PUT /_index_template/logs_template
{
  "index_patterns": ["logs-*"],
  "data_stream": {
    "timestamp_field": "@timestamp"
  }
}
-----------------------------------
// TEST

Start indexing data into the namespace of this composable template:

[source,console]
----
POST /logs-foobar/_doc
{
  "@timestamp": "2050-11-15T14:12:12",
  ...
}
----
// TEST[continued]
// TEST[s/,//]
// TEST[s/\.\.\.//]

Response:

[source,console-result]
--------------------------------------------------
{
    "_shards" : {

>         "total" : 1,

Reviewer: ```suggestion
        "total" : 2,
```

[chapter]
[[data-streams]]
= Data Streams

You can use data streams to index time-based data that's continuously generated.
A data stream groups indices from the same time-based data source.
A data stream tracks its indices, known as _backing indices_, using an ordered
list.

A data stream's backing indices are <<index-hidden,hidden>>.
The most recently created backing index is the

> write index. The other backing indices handle read requests. A data stream only

Reviewer 2: Minor clarification: Read requests are handled by all backing indices, not just the ones that are not the write index.

accepts <<docs-index_,index requests>> with `op_type` set to `create`. To update
or delete specific documents in a data stream, submit a <<docs-delete,delete>>
or <<docs-update,update>> API request to the backing index containing the
document.

To create a data stream, set up a <<indices-templates,composable index
template>> containing:

* Any desired index patterns in the `index_patterns` property. Indices that

>   match these patterns are in the data stream's _namespace_.

Reviewer 2: I find the mention of indices a little confusing and perhaps it's unfortunate that we're using the `index_patterns` parameter here since it's not really about indices. We don't really have (and I don't think we want to introduce) the concept of indices being in a data stream's namespace. While we do have a default naming convention for backing indices, we don't have any constraints on the names of indices that may be backing indices for a data stream, so we couldn't really define a "namespace" for a data stream. I think all that we want to communicate here is that when the pattern in the composable template matches the (not already existing) target of an index request, the corresponding data stream will be created.

* A `data_stream` definition that contains the `timestamp_field` property.

>   The `timestamp_field` should be the primary timestamp field 

Reviewer 2: I don't know how precise we want to be here, but we require the `timestamp_field` property on the template and in every document indexed into the data stream, so "must" may fit better than "should".

   for the data source.

> When you index within a data stream's namespace, {es} automatically creates the

Reviewer 2: I think "data stream's namespace" here is slightly inaccurate. While a composable template may have a pattern such as `logs-*` that many indexing requests may match, indexing requests themselves must specify a single index or data stream name such as `logs-mysql-prod` or the like. In other words, a data stream has a name but not a namespace, and indexing requests must specify the name.

> When you index within a data stream's namespace, {es} automatically creates the
> data stream using this template. You can also manually create a data stream using the

Reviewer: +1 to removing the concept of namespace entirely. I've reworded to use "target" here.
```suggestion
When you index one or more documents to a target matching
the template's name or pattern, {es} automatically creates the corresponding
data stream. You can also manually create a data stream using the
```

<<indices-create-data-stream,create data stream API>>. However, a composable
template for the stream is still required.

You can use the <<indices-rollover-index,rollover API>> to roll a data stream
over to a new index when the current write index meets specified criteria, such
as a maximum age or size. A rollover creates a new backing index and updates the
data stream's list of backing indices. This new index then becomes the stream's
new write index. See <<rollover-data-stream-ex>>.

[discrete]
[[create-data-stream]]
== Create a data stream

Create a composable template with a `data_stream` definition:

[source,console]
-----------------------------------
PUT /_index_template/logs_template
{
  "index_patterns": ["logs-*"],
  "data_stream": {
    "timestamp_field": "@timestamp"
  }
}
-----------------------------------

> Start indexing data into the namespace of this composable template:

Reviewer 2: Same thing about namespace here.

> Start indexing data into the namespace of this composable template:

Reviewer: Reword to use target instead.
```suggestion
Start indexing data to a target matching composable template's wildcard pattern:
```
